core/src/main/java/com/linecorp/armeria/common/logging/ContentPreviewerFactory.java

```diff
@@ -16,216 +16,262 @@
+                    if (contentType.is(candidate)) {
+                        return ContentPreviewer.ofText(maxLength, contentType.charset(defaultCharset));
+                    }
+                }
+            }
+            return ContentPreviewer.disabled();
+        };
     }
 
     /**
-     * Creates a new instance of {@link ContentPreviewerFactory} for creating a {@link ContentPreviewer}.
-     * The previewer produces the text with the maximum {@code length} limit
-     * if the content type of a request/response meets
-     * any of the following conditions:
-     * <ul>
-     *     <li>when it matches {@code text/*} or {@code application/x-www-form-urlencoded}</li>
-     *     <li>when its charset has been specified</li>
-     *     <li>when its subtype is {@code "xml"} or {@code "json"}</li>
-     *     <li>when its subtype ends with {@code "+xml"} or {@code "+json"}</li>
-     * </ul>
-     * @param length the maximum length of the preview.
+     * Returns a new {@link ContentPreviewerFactory} which creates a {@link ContentPreviewer} using the
+     * specified {@link Function}.
+     *
+     * @param maxLength the maximum length of the preview
+     * @param producer the {@link Function} to produce the preview
      */
-    static ContentPreviewerFactory ofText(int length) {
-        return ofText(length, ArmeriaHttpUtil.HTTP_DEFAULT_CONTENT_CHARSET);
+    static ContentPreviewerFactory ofBinary(int maxLength, Function<? super ByteBuf, String> producer) {
+        requireNonNull(producer, "producer");
+        return ofBinary(maxLength, (headers, byteBuf) -> producer.apply(byteBuf));
     }
 
     /**
-     * A dummy {@link ContentPreviewerFactory} which returns {@link ContentPreviewer#disabled()}.
+     * Returns a new {@link ContentPreviewerFactory} which creates a {@link ContentPreviewer} using the
+     * specified {@link Function}.
+     *
+     * @param maxLength the maximum length of the preview
+     * @param producer the {@link Function} to produce the preview
      */
-    static ContentPreviewerFactory disabled() {
-        return NoopContentPreviewerFactory.INSTANCE;
+    static ContentPreviewerFactory ofBinary(int maxLength,
+                                            BiFunction<? super HttpHeaders, ? super ByteBuf, String> producer) {
+        checkArgument(maxLength > 0, "maxLength : %d (expected: > 0)", maxLength);
+        requireNonNull(producer, "producer");
+        return (ctx, headers) -> ContentPreviewer.ofBinary(maxLength, producer, headers);
     }
 
     /**
-     * Returns a {@link ContentPreviewer}, given {@link RequestContext} and {@link HttpHeaders}.
+     * Returns a newly-created {@link ContentPreviewer} with the given {@link RequestContext} and
+     * {@link HttpHeaders}. Note that the returned {@link ContentPreviewer} can be
+     * {@link ContentPreviewer#disabled()}.
      */
     ContentPreviewer get(RequestContext ctx, HttpHeaders headers);
```

Comment:
How about splitting this method into two:

```java
ContentPreviewer requestContentPreviewer(ctx, RequestHeaders headers);
ContentPreviewer responseContentPreviewer(ctx, RequestHeaders reqHeaders, ResponseHeaders resHeaders);
```

I guess the behavior of response previewer might be affected by a certain request header, e.g. `x-enable-content-preview: 1`.

Reply:
That's a good idea. I need more refactoring. 😉